File: ue-sdk-android/src/main/java/et/restlink/uesdk/android/SessionTupleClient.java
/*
 * Silent Auth UE SDK (Android) — Restlink (Ethiopia).
 * Device-side session-tuple poster. Java 8 / Android minSdk 24.
 * R&D only — never production.
 *
 * Copyright (c) 2026 Tran Nhan (nhanth87). All rights reserved.
 */

package et.restlink.uesdk.android;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

/**
 * Posts a {@link TupleSnapshot} to the SAS {@code POST /session-tuple}
 * endpoint over {@link HttpURLConnection}.
 *
 * <p>Body shape mirrors the server DTO ({@code srcIp, srcPort, ts, msisdn,
 * imsi}); null fields are omitted. Run {@link #post} off the main thread —
 * this is blocking network I/O.</p>
 */
public final class SessionTupleClient {

    private final int connectTimeoutMs;
    private final int readTimeoutMs;

    public SessionTupleClient() {
        this(3000, 3000);
    }

    public SessionTupleClient(int connectTimeoutMs, int readTimeoutMs) {
        this.connectTimeoutMs = connectTimeoutMs;
        this.readTimeoutMs = readTimeoutMs;
    }

    /**
     * Returns the HTTP status code; throws {@link IOException} on transport failure.
     */
    public int post(String sasBaseUrl, TupleSnapshot snapshot,
                    String apiKeyNullable) throws IOException {
        URL url = new URL(trimTrailingSlash(sasBaseUrl) + "/session-tuple");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setConnectTimeout(connectTimeoutMs);
        conn.setReadTimeout(readTimeoutMs);
        conn.setDoOutput(true);
        conn.setRequestProperty("Content-Type", "application/json");
        if (apiKeyNullable != null && apiKeyNullable.trim().length() > 0) {
            conn.setRequestProperty("X-Api-Key", apiKeyNullable);
        }
        byte[] body = body(snapshot).getBytes(StandardCharsets.UTF_8);
        try (OutputStream out = conn.getOutputStream()) {
            out.write(body);
        }
        int status = conn.getResponseCode();
        InputStream errorStream = conn.getErrorStream();
        if (errorStream != null) {
            try {
                errorStream.close();
            } catch (IOException ignored) {
                // best-effort drain
            }
        }
        conn.disconnect();
        return status;
    }

    static String trimTrailingSlash(String baseUrl) {
        return baseUrl != null && baseUrl.endsWith("/") && baseUrl.length() > 1
                ? baseUrl.substring(0, baseUrl.length() - 1)
                : baseUrl;
    }

    static String body(TupleSnapshot snapshot) {
        StringBuilder sb = new StringBuilder(128);
        sb.append('{');
        boolean first = true;
        if (snapshot.srcIp() != null) {
            sb.append("\"srcIp\":").append(Json.quote(snapshot.srcIp()));
            first = false;
        }
        if (snapshot.srcPort() != null) {
            appendComma(sb, first);
            sb.append("\"srcPort\":").append(snapshot.srcPort().intValue());
            first = false;
        }
        appendComma(sb, first);
        sb.append("\"ts\":").append(snapshot.ts());
        first = false;
        if (snapshot.claimedMsisdn() != null) {
            appendComma(sb, first);
            sb.append("\"msisdn\":").append(Json.quote(snapshot.claimedMsisdn()));
            first = false;
        }
        if (snapshot.imsi() != null) {
            appendComma(sb, first);
            sb.append("\"imsi\":").append(Json.quote(snapshot.imsi()));
        }
        sb.append('}');
        return sb.toString();
    }

    private static void appendComma(StringBuilder sb, boolean first) {
        if (!first) {
            sb.append(',');
        }
    }
}

File: ue-sdk-ios/Sources/UESDK/SessionTupleClient.swift
/*
 * Silent Auth UE SDK (iOS) — Restlink (Ethiopia).
 * Device-side session-tuple poster. R&D only — never production.
 *
 * Copyright (c) 2026 Tran Nhan (nhanth87). All rights reserved.
 */

import Foundation

/// Posts a `TupleSnapshot` to the SAS `POST /session-tuple` endpoint over
/// `URLSession`.
///
/// Body shape mirrors the server DTO (`srcIp`, `srcPort`, `ts`, `msisdn`,
/// `imsi`); null fields are omitted. Contains no authentication logic;
/// approval always happens server-side (fail-closed FSM in the SAS).
public final class SessionTupleClient {

    private let baseURL: URL
    private let apiKey: String?
    private let session: URLSession
    private let bearer: CellularBearer?
    private let requirement: CellularBearer.Requirement

    /// - Parameters:
    ///   - baseUrl: SAS base URL; a single trailing slash is trimmed.
    ///   - apiKey: sent as `X-Api-Key` only when non-nil and non-empty.
    ///   - timeout: request/resource timeout in seconds (SAS contract: 3).
    ///   - configuration: injectable for tests (URLProtocol stubbing). For a
    ///     real cellular attempt pass
    ///     `CellularBearer.cellularSessionConfiguration(timeout:)`.
    ///   - bearer: observed-bearer gate; nil means "do not check" (lab only).
    ///   - requirement: how strictly cellular is demanded.
    public init(baseUrl: String,
                apiKey: String? = nil,
                timeout: TimeInterval = 3.0,
                configuration: URLSessionConfiguration = .default,
                bearer: CellularBearer? = nil,
                requirement: CellularBearer.Requirement = .any) {
        var trimmed = baseUrl
        if trimmed.count > 1 && trimmed.hasSuffix("/") {
            trimmed.removeLast()
        }
        guard let url = URL(string: trimmed) else {
            preconditionFailure("Invalid SAS base URL: \(baseUrl)")
        }
        self.baseURL = url
        self.apiKey = apiKey
        self.bearer = bearer
        self.requirement = requirement
        configuration.timeoutIntervalForRequest = timeout
        configuration.timeoutIntervalForResource = timeout
        self.session = URLSession(configuration: configuration)
    }

    deinit {
        session.finishTasksAndInvalidate()
    }

    /// Sends the snapshot; returns the HTTP status code.
    ///
    /// - Throws: `CellularBearer.Unavailable` when the gate cannot confirm the
    ///   demanded bearer — before any request leaves the device, so the SAS
    ///   never sees a Wi-Fi address presented as cellular.
    @discardableResult
    public func send(msisdn: String? = nil) async throws -> Int {
        var declared: AccessTech?
        if let bearer = bearer {
            try bearer.check(requirement)
            let observed = bearer.accessTech()
            declared = observed == .unknown ? nil : observed
        }

        var request = URLRequest(url: baseURL.appendingPathComponent("session-tuple"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let declared = declared {
            request.setValue(declared.rawValue, forHTTPHeaderField: "X-Sas-Access-Tech")
        }
        if let apiKey = apiKey, !apiKey.isEmpty {
            request.setValue(apiKey, forHTTPHeaderField: "X-Api-Key")
        }
        request.httpBody = try JSONEncoder().encode(
            TupleSnapshot(msisdn: msisdn, accessTech: declared))

        let (_, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        return http.statusCode
    }
}

